network/bridge_port_test.go

```diff
@@ -3,4 +3,162 @@
 // Copyright (c) 2022-2023 Dell Inc, or its subsidiaries.
 
 // Package network implements the go library for OPI to be used to establish networking
-package network
+package network_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/opiproject/godpu/mocks"
+	_go "github.com/opiproject/opi-api/network/evpn-gw/v1alpha1/gen/go"
+	emptypb "google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+func TestCreateBridgePortSuccess(t *testing.T) {
+	mockEvpnClient := &mocks.EvpnClient{}
+
+	expectedBridgePort := &_go.BridgePort{} // Create your expected response
+	mockEvpnClient.On("CreateBridgePort", mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything).
+		Return(expectedBridgePort, nil)
+
+	resultBridgePort, err := mockEvpnClient.CreateBridgePort(context.TODO(), "bp1", "00:11:22:AA:BB:CC", "access", []string{"lb1", "lb2"})
+
+	assert.NoError(t, err)
+	assert.Equal(t, expectedBridgePort, resultBridgePort)
+	mockEvpnClient.AssertExpectations(t)
+}
+
+func TestCreateBridgePortError(t *testing.T) {
+	mockEvpnClient := &mocks.EvpnClient{}
+
+	expectedError := errors.New("mocked error")
+	mockEvpnClient.On("CreateBridgePort", mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything).
+		Return(nil, expectedError)
+
+	resultBridgePort, err := mockEvpnClient.CreateBridgePort(context.TODO(), "bp1", "00:11:22:AA:BB:CC", "access", []string{"lb1", "lb2"})
+
+	assert.Error(t, err)
+	assert.Nil(t, resultBridgePort)
+	assert.Equal(t, expectedError, err)
+	mockEvpnClient.AssertExpectations(t)
+}
+
+func TestDeleteBridgePortSuccess(t *testing.T) {
+	mockEvpnClient := &mocks.EvpnClient{}
+
+	expectedResponse := &emptypb.Empty{} // Create your expected response
+	mockEvpnClient.On("DeleteBridgePort", mock.Anything, mock.Anything, mock.Anything).
+		Return(expectedResponse, nil)
+
+	resultResponse, err := mockEvpnClient.DeleteBridgePort(context.TODO(), "bp1", true)
+
+	assert.NoError(t, err)
+	assert.Equal(t, expectedResponse, resultResponse)
+	mockEvpnClient.AssertExpectations(t)
+}
+
+func TestDeleteBridgePortError(t *testing.T) {
+	mockEvpnClient := &mocks.EvpnClient{}
+
+	expectedError := errors.New("mocked error")
+	mockEvpnClient.On("DeleteBridgePort", mock.Anything, mock.Anything, mock.Anything).
+		Return(nil, expectedError)
+
+	resultResponse, err := mockEvpnClient.DeleteBridgePort(context.TODO(), "bp1", true)
```

Comment:
@atulpatel261194 A bit late, but why do you test the generated mock here in these tests? Testing of mocks adds little value to our test suite.
I think we should test the implementation(evpnClientImpl) calls a corresponding method in pb.BridgePortServiceClient instead...

Reply:
@artek-koltun testing the generated mocks protects the changes in .proto files.
i believe we can't test the real api's which create, delete and update different port, bridge, svi and vrf. if there is any way i can test these api's directly Please let me know.

Reply:
There is actually a way. You can use dependency injection to replace the real connection components. The test would look like
```Go
func TestCreateBridgePortSuccess(t *testing.T) {
	wantRequest := &_go.CreateBridgePortRequest{
		BridgePortId: "bp1",
		BridgePort: &pb.BridgePort{
			Spec: &pb.BridgePortSpec{
				MacAddress:     []byte("00:11:22:aa:bb:cc"),
				Ptype:          _go.BridgePortType_ACCESS,
				LogicalBridges: []string{"lb1", "lb2"},
			},
		},
	}
	wantResponse := wantRequest.BridgePort

	mockClient := mocks.NewBridgePortServiceClient(t)
	mockClient.EXPECT().CreateBridgePort(mock.Anything, wantRequest).
		Return(proto.Clone(wantResponse).(*pb.BridgePort), nil)

	connClosed := false
	mockConn := mocks.NewConnector(t)
	mockConn.EXPECT().NewConn().Return(
		&grpc.ClientConn{},
		func() { connClosed = true },
		nil,
	)

	c, _ := network.NewBridgePortWithArgs(
		mockConn,
		func(grpc.ClientConnInterface) pb.BridgePortServiceClient {
			return mockClient
		},
	)
	
	response, err := c.CreateBridgePort(
		context.Background(),
		"bp1", "00:11:22:aa:bb:cc", "access", []string{"lb1", "lb2"},
	)

	assert.NoError(t, err)
	assert.True(t, proto.Equal(response, wantResponse))
	assert.Equal(t, true, connClosed)
}
```

[Here](https://github.com/opiproject/godpu/blob/main/storage/nvme_subsystem_test.go) you can see how I tested subsystem with the table approach